src/errors.rs

```diff
@@ -35,66 +35,55 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-use std::{io, ffi, num};
 use std::fmt::{self, Formatter, Display};
+use std::error::Error;
+use TracerState;
 
 #[derive(Debug)]
 pub enum HWTracerError {
-    // Wrapped errors from elsewhere.
-    FFIIntoString(ffi::IntoStringError),
-    FFINul(ffi::NulError),
-    IO(io::Error),
-    NumParseInt(num::ParseIntError),
-    // Our own errors.
-    CFailure,
-    ElfError(String),
-    HardwareSupport(String),
-    InvalidFileName(String),
-    TracerAlreadyStarted,
-    TracerDestroyed,
-    TracerNotStarted,
-    TracingNotPermitted(String),
-}
-
-impl From<ffi::IntoStringError> for HWTracerError {
-    fn from(err: ffi::IntoStringError) -> Self {
-        HWTracerError::FFIIntoString(err)
-    }
-}
+    // Non-fatal errors.
+    HWBufferOverflow,     // The trace buffer being used by the hardware overflowed.
+                          // This is considered a non-fatal error since retrying the tracing may
+                          // succeed.
+    HWSupport(String),    // The hardware doesn't support a required feature. Not fatal for the
+                          // same reason as `TracingPerms`.
+    TracingPerms(String), // Tracing is not permitted. Non-fatal as the consumer could carry on
+                          // without tracing support.
 
-impl From<ffi::NulError> for HWTracerError {
-    fn from(err: ffi::NulError) -> Self {
-        HWTracerError::FFINul(err)
-    }
+    // Fatal errors.
+    CFailure,                 // Something went wrong in C code.
+                              // ^ XXX will be replaced with an errno mechanism.
+    TracerState(TracerState), // The tracer is in the wrong state to do the requested task.
+    FatalError(Box<Error>),   // All other errors. Boxed so that individual backends can choose
```

Comment:
Should we call this `UnknownError` or `CustomError` or similar?

Reply:
`io::Error` in the Rust std lib uses both `Other` and `Custom`. It's a bit more complicated there, as they separate the error type from it's representation, but it all boils down to:

```
enum Repr {
    Os(i32),
    Simple(ErrorKind),
    Custom(Box<Custom>),
}
```

Which is amusingly recursive.

https://doc.rust-lang.org/src/std/io/error.rs.html#77

The last push uses `Custom` for our "other error type".